### PR TITLE
change: components apply after all generators and transformers applied

### PR DESCRIPTION
**Pull request by koba1t**
fix: https://github.com/kubernetes-sigs/kustomize/issues/5141

### CODE REVIEW

**Comment by natasha41575** [comment]
/approve

@annasong20, when you give this PR a review could you also build off the branch to be a second verifier that it fixes https://github.com/kubernetes-sigs/kustomize/issues/5141?

File: api/internal/target/kusttarget.go [hunk 1]
```diff
@@ -238,6 +234,11 @@ func (kt *KustTarget) accumulateTarget(ra *accumulator.ResAccumulator) (
 		return nil, errors.WrapPrefixf(
 			err, "merging vars %v", kt.kustomization.Vars)
 	}
+	// components are accumulated last.
```
Comment (annasong20) [inline api/internal/target/kusttarget.go:237]: In this comment, can we explain why "components are accumulated last"?

File: api/krusty/component_test.go [hunk 2]
```diff
@@ -657,3 +657,63 @@ components:
 		})
 	}
 }
+
+func TestOrderOfAccumulatedComponent(t *testing.T) {
+	tests := map[string]struct {
+		input          []FileGen
+		expectedOutput string
+	}{
+		"a": {
```
Comment (annasong20) [inline api/krusty/component_test.go:666]: Can we rename the test to something more descriptive? For example, "after generators".

**Review by annasong20 — CHANGES_REQUESTED** [review]
Only readability comments. Great spacing and `go.work.sum` fixes.

**Comment by annasong20** [comment]
> @annasong20, when you give this PR a review could you also build off the branch to be a second verifier that it fixes #5141?

Fixes the issue!

File: api/krusty/component_test.go (hunk 2 above)
Comment (koba1t) [reply api/krusty/component_test.go:666]: Oh!
Sorry, I forgot to fix the test name.

Comment (koba1t) [reply api/krusty/component_test.go]: I fixed at https://github.com/kubernetes-sigs/kustomize/pull/5170/commits/0c38baae5e83ce8d377dac064f82d5130b4a80f1.

**Comment by koba1t** [comment]
Hi @annasong20 

I have done a fix.
Could you recheck it?

File: api/internal/target/kusttarget.go [hunk 3]
```diff
@@ -238,6 +234,12 @@ func (kt *KustTarget) accumulateTarget(ra *accumulator.ResAccumulator) (
...
+	// components are expected to execute after reading resources, adding generators, and applying transformers.
+	ra, err = kt.accumulateComponents(ra, kt.kustomization.Components)
```
[the remark this answers is not in the document]
Comment (annasong20) [reply api/internal/target/kusttarget.go:239]: @koba1t I'll review the PR again after this change.

[the remark this answers is not in the document]
Comment (koba1t) [reply api/internal/target/kusttarget.go]: @annasong20 
I moved the code and added a test case. https://github.com/kubernetes-sigs/kustomize/pull/5170/commits/1d0a14d9d5ac2477aeb9de1e5b0c5865bd6a1141
Could you recheck it?

**Comment by ephesused** [comment]
FYI, we are seeing some kustomize build failures associated with this commit. I'm working to isolate the source and create a simple test case.

**Comment by ephesused** [comment]
I believe the failures we're seeing are expected given this PR's description, but if someone is willing to confirm, I'd appreciate it. My understanding is that our current pattern (generate a configmap in the component and merge it in the top-level kustomization.yaml) will need to be inverted (generate a configmap in the top-level kustomization.yaml and merge it in the component). Is that accurate?

kustomization.yaml:

```yaml
components:
- component-example
configMapGenerator:
- files:
  - datafile.txt
  name: cm-for-datafile
  behavior: merge
```

component-example/kustomization.yaml:

```yaml
kind: Component

generators:
- generators.yaml
```

component-example/generators.yaml:

```yaml
apiVersion: builtin
kind: ConfigMapGenerator
metadata:
  name: cm-for-datafile
labels:
  app.kubernetes.io/name: cm-for-datafile
```

datafile.txt is empty.

Results from kustomize builds before and after this commit:

```
[root@e7265ac0feda pr5170]# ../../kustomize-execs/branch-6ce0bf39/kustomize build .
apiVersion: v1
data:
  datafile.txt: ""
kind: ConfigMap
metadata:
  name: cm-for-datafile-hbfc9tkcd5
[root@e7265ac0feda pr5170]# ../../kustomize-execs/branch-0add0f95/kustomize build .
Error: merging from generator &{0xc0004381a0 <nil>}: id resid.ResId{Gvk:resid.Gvk{Group:"", Version:"v1", Kind:"ConfigMap", isClusterScoped:false}, Name:"cm-for-datafile", Namespace:""} does not exist; cannot merge or replace
[root@e7265ac0feda pr5170]#
```

**Comment by annasong20** [comment]
> My understanding is that our current pattern (generate a configmap in the component and merge it in the top-level kustomization.yaml) will need to be inverted (generate a configmap in the top-level kustomization.yaml and merge it in the component). Is that accurate?

@ephesused Yes, I believe that should solve the problem. Thank you for the detailing your setup and feel free to leave any feedback for this change at #5141.

**Comment by seh** [comment]
@koba1t, the title of this PR is confusing: It doesn't match the comments in the code, and doesn't match the decisions reached in #5141. As far as I can tell, we're applying components _before_ transformers. Is that correct?

**Comment by koba1t** [comment]
/retitle components are applied after generators, before transformers

**Comment by koba1t** [comment]
@seh

That's right. Thanks!

**Pull request merged**